### DISCUSSION TITLE
Source Generators need to be able to call other source generators

### DISCUSSION THREAD

**Discussion by YairHalberstadt** [dotnet/roslyn#48358] [Ideas]
There are plans to push for source generators to be used for all sorts of scenarios that are currently done at runtime. For example:

1. The runtime is looking into using SourceGenerators to generate interop stubs: https://github.com/dotnet/runtime/blob/master/docs/design/features/source-generator-pinvokes.md

2. There's a push to get serialization frameworks to use Source Generators to help with trimming among other scenarios: https://devblogs.microsoft.com/dotnet/app-trimming-in-net-5/

I'm sure there are many more - for example Asp.Net is definitely thinking about how it can use Source Generators to wire things up.

However at the moment source generators can't use other source generators. What this means is that it will be impossible for a source generator to generate pinvoke code, asp.net code, or code using serialization, since the relevant source generators which are meant to process these wont see the generated files (although they can always use the runtime/reflection option as a fallback). This seems like a pretty critical limitation.

In an ideal world it would be possible to create a graph of exactly which source generators rely on which other ones. However in practice this means creating potentially as many compilations as there are source generators which may lead to serious performance problems: https://discord.com/channels/732297728826277939/735233259763400715/762740254921719848. There's also no obvious, reliable way to specify this graph.

More pragmatically I think it would be sufficient to run source generators in 2 passes. By default all source generators run in the second pass, and cannot use other source generators. However a SourceGenerator can opt in to running in the first pass. This will mean files it generates will be seen by the second pass. So either your generator can use other generators, or it can be used by other generators, but not both.

That way low level source generators which expect to be used by other source generators will make sure not to rely on any other source generator (even if that leads to a lot of extra work). Other source generators can then freely use these if they want.

**Comment by jasonmalinowski** [comment]
We've so far intentionally tried to avoid this because of all of the problems you've mentioned. :smile:

> By default all source generators run in the second pass, and cannot use other source generators. However a SourceGenerator can opt in to running in the first pass.

I worry about this a bit though: a serialization library needs to run "after" other any other thing trying to generate a type because it might have to serialize that type. So that implicitly means anything generating a type has to be in the _first_ pass, not second pass. And the challenge here is if somebody writes a generator and didn't opt themselves into the first pass, then the user is now broken when they try to use both...

**Reply by YairHalberstadt** [reply]
Surely a serialisation library wouldn't have to generate serialisization code for everything - that would be super inefficient. It would have to be opt in only.

Also only a code generator generating exchange types is likely to require serialisation of those types.

**Reply by jasonmalinowski** [reply]
I think the expectation is some serializers can "opt in" by simply seeing uses: if you see some code calling `Serializer.Serialize<MyType>` then it knows it needs to generate the logic for serializing MyType; if MyType came from another generator then it'll need to see that type.

**Reply by YairHalberstadt** [reply]
Well in that case the Json Serializer will give an error because it can't see `MyType`.

There could be an option in the csproj to change which pass a generator runs in from the default. In the relatively rare case where A depends on B depends on C you would be stuck, but I think worrying about that would be letting perfect be the enemy of good.

**Comment by stefanloerwald** [comment]
I'm convinced sequencing is something absolutely crucial, also w.r.t. breaking changes: for example, blazor/razor currently uses MSBuild-based source generation (a razor file is converted to a csharp file which then gets compiled normally). This is something that currently doesn't make use of roslyn-SGs, but in the future might very well do so. However, we can currently write source generators that use the MSBuild-SG-generated code, e.g. to inspect parameters of components (`razor` files are components, which translate to classes with properties annotated with the `[Parameter]` attribute). If the generator were now to be replaced by a roslyn-SG, that now breaks every roslyn-SG that previously relied on the generated code. Requiring every SG to look through `razor` files too would duplicate the effort (and would be error-prone), so the only good choice for this problem is to have a clear sequence "razor-SG" before "my-SG".

My gut feeling says that generating a dependency graph (enforced to be cycle-free) should suffice to get rid of most problems.

**Reply by YairHalberstadt** [reply]
> My gut feeling says that generating a dependency graph (enforced to be cycle-free) should suffice to get rid of most problems.

Generating a dependency graph would definitely be the ideal solution, but there's 2 problems:

1. How do you manage the performance problems inherent in making multiple compilations? 

I think we need to see more evidence for how much of an impact this has in practice: I imagine in most cases, the dependency graph will ending up only requiring 2, 3, or maybe even 4 unique compilations. Creating a new compilation is basically free - what costs is any symbols generated as part of using the compilation in the source generator get thrown away in the next phase, so it's not like 4 compilation passes is 4 times as slow as just 1.

2. How do you specify the dependency graph?

Doing something like `DependsOn(typeof(SomeSourceGenerator))` wont work since the actual generator class is an implementation detail, and is often internal. Also that would require packaging dependent source generators with the generator that uses them.

**Reply by stefanloerwald** [reply]
> 2. How do you specify the dependency graph?
> 
> Doing something like `DependsOn(typeof(SomeSourceGenerator))` wont work since the actual generator class is an implementation detail, and is often internal. Also that would require packaging dependent source generators with the generator that uses them.

I'd imagine specifying dependencies should probably be done using csproj: as a developer, I can specify a list of SGs that I know are always required to go before my SG, but the user might have a few more. So the package will come with a few default dependencies as defined by the package author, and if that's not sufficient, the user can add some more.

**Reply by YairHalberstadt** [reply]
> I'd imagine specifying dependencies should probably be done using csproj: as a developer, I can specify a list of SGs that I know are always required to go before my SG, but the user might have a few more.

You mean by referencing the assembly/nuget package in the csproj? That makes a lot of sense.

**Reply by YairHalberstadt** [reply]
To give an idea how this would work:

I have 5 generators:

1. Json generator
2. Interop generator
3. Generator A (depends on json)
4. Generator B (depends on interop)
5. Generator C (depends on json and interop)

Then we do as follows

1. Create an initial compilation InitialComp
2. Run generators A, B and C with the initial compilation, to generate AFiles, BFiles and CFiles.
3. Create a compilation InitialACComp consisting of InitialComp+ AFiles + CFiles
4. Run json generator with InitialACComp, to generate JsonFiles
5. Create a compilation InitialBCComp consisting of InitialComp+ BFiles + CFiles
6. Run interop generator with InitialBCComp to generate InteropFiles
7. Create a final compilation consisting of Initial + AFiles + BFiles + CFiles + JsonFiles + InteropFiles

Here a total of 4 compilations are needed.
As you can see the maximum number of compilations is equivalent to the number of generators + 1, since in the worst case you need a new compilation for every generator.

**Reply by stefanloerwald** [reply]
> To give an idea how this would work:
> 
> I have 5 generators:
> 
> 1. Json generator
> 2. Interop generator
> 3. Generator A (depends on json)
> 4. Generator B (depends on interop)
> 5. Generator C (depends on json and interop)
> 
> Then we do as follows
> 
> 1. Create an initial compilation InitialComp
> 2. Run generators A, B and C with the initial compilation, to generate AFiles, BFiles and CFiles.

Why would A, B, and C run before their dependency completed?

If I understand your dependencies right you have

```none
IOp ---> B
     \
       > C
     /
Json ---> A 
```
or did you mean it the exact other way round?

In my head the terminology "A depends on the JSON-SG" would mean "A should only run once it can see the output of the JSON-SG", which would make runnign the A-SG before the JSON-SG rather pointless.

**Reply by YairHalberstadt** [reply]
I meant the other way round. The way I'm thinking of it is, in most cases a source generator will generate code that it will want another generator to consume. The second generator is agnostic to where the code it sees comes from. Therefore if a depends on b, a runs first, and then b fixes the code generated by a.

However I guess you would want to be able to specify it both ways - this generator runs before generator X, and after generator y.

**Reply by stefanloerwald** [reply]
Agreed. Maybe something similar to what MSBuild does with `AfterTargets`/`BeforeTargets`?

**Reply by svick** [reply]
The micro-optimzer in me thinks this could be improved as follows:

1. Create an initial compilation InitialComp.
2. Run generators A, B and C with the initial compilation, to generate AFiles, BFiles and CFiles.
3. Create a compilation InitialABCCComp consisting of InitialComp+ AFiles + BFiles + CFiles.
4. Run json generator and interop generator with InitialABCComp, to generate JsonFiles and InteropFiles.
5. Create a final compilation consisting of Initial + AFiles + BFiles + CFiles + JsonFiles + InteropFiles.

This still satisfies all dependencies and results in fewer compilations being created. (The actual number depends on the length of the critical path in the graph.)

Though maybe doing it this way is not a good idea, because it means which set of files each generator sees would be kind of unpredictable.

**Comment by dallmair** [comment]
Just faced this in a custom source generator, which generates a custom `JsonSerializerContext` and expected it to be picked up by the STJ source generator -- wrote the code although I knew source generators operate on the user's compilation unit and cannot see each other's generated source.

So definitely 👍 for *some* solution in this space.

**Reply by domportera** [reply]
im trying to do the exact same thing and hitting this wall over and over. did you ever come up with a solution?

**Comment by msedi** [comment]
If this would be possible, would it then also be possible to let source generators modify the code and not just add code?

**Reply by canton7** [reply]
This has been discussed multiple times. The mechanism being introduced for this is Interceptors: https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-12#interceptors

**Reply by msedi** [reply]
@canton7: I know about the interceptors. But I think they require valid code.
For my current need I have invalid code, a Unit class that represents a physical value with a unit on it. Currently I have to write some constructors that need to be parsed at runtime and only fail there. So this is how the code currently looks like

```
Unit a = "12.0mm";
Unit b = new Unit(12.0, "mm");
```

in both cases it will fail if something was written with a wrong physical unit. So i need something where I can simply write

`Unit c = 12.0mm;`

analyze the code during source generation and replace the code to

`Unit d = new Unit(12.0, Units.mm);

But this is not what Interceptors can do, right?

**Reply by 333fred** [reply]
> For my current need I have invalid code

We are not interested in enabling users to create custom dialects of C#. There is no guarantees around how invalid syntax is parsed, especially in the face of new language features that may change various implementation details.

**Reply by fjmorel** [reply]
> Currently I have to write some constructors that need to be parsed at runtime and only fail there

You could also write a Roslyn analyzer to make sure the strings fit a certain format and add a diagnostic (and maybe even a code fix!) to bad strings.

**Reply by msedi** [reply]
@fjmorel. True, but its not about diagnostics only, its more also about performance. The upper example Unit a = "12.0mm"; causes an tremendous overhead since I need to parse a string, split it and look up the necessary exponents.

**Reply by msedi** [reply]
@333fred: I agree, and I don't want that either. I'm glad that you didn't bring C++ macros into the language. I'm only searching for a way to make physical computations easy to read and performant. The current approach is not a solution. So the question is if interceptors git be a way to do something like the above.

**Reply by 333fred** [reply]
@msedi then I'm confused. You specifically asked to be able to write `Unit c = 12.0mm;` and have that be interpreted by a rewriting generator. That is _exactly_ what I'm talking about when I say that we are not interested in allowing creation of custom dialects. Users need to write valid C# code to start with.

**Reply by CyrusNajmabadi** [reply]
`12.0.mm();` seems to be something you coudl trivially do.  With 'extension everything' coming in C# vNext, you could do it as `12.0.mm`.

**Reply by msedi** [reply]
@CyrusNajmabadi. Thats an approach we currently pursue with extension methods `12.0.mm()`. Though it gets ugly when you have something like [1/sr*keV*mAs]. This would mean you have a few mathematical operations on the data only for initialization and I have no clue how to write that with extension methods/properties.

That would mean
12.0 / 1.sr / 1.keV / 1.mAs which completely obscures the unit

**Reply by CyrusNajmabadi** [reply]
Make helpers which combine a lot of the units together.

**Reply by msedi** [reply]
It should be a general type which means I need to combine all possible combinations of all possible units that can exist. You can see a list of these quantities here: [wiki](https://en.wikipedia.org/wiki/List_of_physical_quantities). With the list multiple permutations/combinations are possible and it would be almost impossible to precreate all of these.

I understand that this is a special case and that there should not be any dialect as @333fred said. I'm just trying to find a native way to describe units which makes keeps code readable.

**Reply by canton7** [reply]
Can you please take this to another discussion? This discussion is about source generators calling other source generators, and a lot of people are being emailed every detail about your units problem, which is not relevant to us. Thanks!

**Reply by CyrusNajmabadi** [reply]
> which means I need to combine all possible combinations of all possible units that can exis

You only need to define extensions for cases you actually end up using.  Normally this would fall out from the math operations you are doing.  In the rarer cases where you are defining an initial value, you can just have extensions to help there.

**Reply by msedi** [reply]
We are in a physics department ;-) so I cannot forsee what units we have, we are dealing with all (at least too many) of them ;-)
But I looked at other solutions from Python, C++, etc. all of them need to solve it that way lacking a better idea (or language support).

**Reply by CyrusNajmabadi** [reply]
Add when you run into something new you need. You'd have to do that anyway with a new dialect of the language.

**Comment by dowhilefor** [comment]
Also stumbled upon this. Implementing a sort of RPC framework. For that im putting the arguments of the methods into MemoryPack classes and add some boiler plate using a source generator. Unfortunately the MemoryPack source generator does not see my generated Argument files and therefore it can't generate it. Now i'm contemplating if i should write a manualy code parser as a pre-build step, that generates the files, adds it to the prj and then the MemoryPack generator can do its work.

**Comment by mwadams** [comment]
I have just hit this issue with the source generator in https://github.com/corvus-dotnet/corvus.jsonschema

This emits code for serialization/validation using JSON Schema files as the basis. 

When you run the command line tool, the code it emits uses the `System.Text.RegularExpressions` SourceGenerator.

This will not compile, so we have to fall back on runtime creation of Regular Expressions, which (I assume) will negatively impact start-up time.

**Comment by thomhurst** [comment]
I'm the author of TUnit, a testing framework that uses source generators.

Recently a user tried source-generating their tests, but then of course TUnit also uses source generators, so it can't actually see their tests to discover and register them.

So a good example where it'd be nice to allow source generators to see generated code.

**Reply by CyrusNajmabadi** [reply]
@thomhurst No design has been presented that seems like it can adequately solve this.

**Comment by ChristianSauer** [comment]
We've also run into this, see: https://github.com/codecentric/net_automatic_interface/issues/73
My gut feeling says that this needs to be solved soonish, SGs are taking off in dotnet and this a really bad wall to run into

**Comment by Cubody** [comment]
Same here. I made enum attributes metadata generator and simple mapper generator, that generates code from MapTo() invocations. And I can't use MapTo() with my enum metadata classes.

I can see 2 possible workarounds for now.
Firstly I'll describe abstract example(SG - Source Generator):
SG A generates types in the MyProject assembly
SG B generates some code into the MyProject assembly based on its types, but ignores the SG B types.

Possible workarounds:
1. Make backward compatibility packages that call SG B code for SG A, and set a public method for calling SG B. I can't imagine how many packages in packages it can grow into...
2. Make separate csproj, where SG A generates the code, and refer it to another csrpoj, where SG B is connected (like okay, but it seems to be an inconvenience from a bunch of domain packages for each individual SG case)

P.S. I understand that there are difficulties with implementing "visibility of generated code", but without SG chaining, we seem to be deprived of arms and legs. Is it low priority?

**Comment by akobr** [comment]
I'm having similar issues. The source generators are fantastic, and I use them in many places, but this is going to be a growing problem. More and more tools are based on generators to avoid reflection and other pitfalls. However, when we start generating input for them, we face the issue of not being able to chain generators. Tools like **TUnit** or now my own problem when I'm generating grains for **Orleans**.

**Reply by akobr** [reply]
I managed to find a workaround, but I don't like it. To have **Project A** where I run my generator and `EmitCompilerGeneratedFiles` set to `true`. Then I have another **Project B** (second compilation) where I reference Orleans SDK (their generators) and add the emited code from project A as `Compile` content. I'm referencing and working with the Project B, in that case I have my own code generated and all the generated stuff from Orleans. It is nusty and complicated but it works.

**Comment by domportera** [comment]
I managed to do some unholy reflection nonsense to get this to work,  though i think i only got the non-incremental generator working. a proper api proposal for chaining is definitely necessary
https://github.com/domportera/NetJsonAOT

**Comment by ZzZombo** [comment]
What stops us from getting a new pipeline feature like `WithGenerator<IGenerator>(instance)`?

**Reply by CyrusNajmabadi** [reply]
@ZzZombo it's unclear what that means.

**Reply by willnationsdev** [reply]
I assume this is them suggesting that the code for GeneratorA explicitly calls the method above with a generic type reference for GeneratorB, thereby *somehow* marking GeneratorB as a dependency of GeneratorA.

But regardless, I don't think a pattern like that would be effective. More likely it would be some sort of MSBuild configuration thing in the .csproj file rather than something you call directly in the source generator's code.

**Reply by ChristianSauer** [reply]
@willnationsdev I would expect something like this:
    <PackageReference Include="AutomaticInterface">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <CompilationPass>1</CompilationPass>
    </PackageReference>
    CompilationPass would define the order. This is manual work, yes - but much better than "does not work at all"

**Comment by alexnayd-dev** [comment]
Plus one :+1: 
Trying to write log-proxy generator which relies on Microsoft.Extensions.Logging.LoggerMessageAttribute, including complex objects logging with redaction (Microsoft.Extensions.Compliance.Redaction)
